tests: updata the double list testcase

Zhaoningx opened the pull request:
add some testcases to verify that can be embeddedable in user structure,
verify dlist "for each" style.

Signed-off-by: Ningx Zhao <ningx.zhao@intel.com>

Commit:
tests: updata the double list testcase

Add some testcases to:
Verify an user defined structure contains dlist node works,
Verify dlist "for each" style APIs work.

Signed-off-by: Ningx Zhao <ningx.zhao@intel.com>

andyross reviewed (commented):
The code looks OK, but I'm wondering if this should move to a benchmark or something?  Adding CPU-bound tests to CI is very expensive. 

It's not a bad idea to prove the run time of algorithms, but... do we really need to do that to show the operation on a doubly-linked list is constant time?  You can just as easily read the code and verify that it has no loops/recursion!

Zhaoningx commented:
Hi @andyross , I totally agree with you, it too expensive to use the statistical method to verify the double-list operations running in constant time. I have moved the code to a benchmark, and I have changed code a lot.

There are something confused me, I think double list should be a unit test. But now, we changed the path into a benchmark, can it be a unit test or should depend on the specific platform? In my opinion, I think double list is just a C data-structure, so which platform to test the performance of double-list, it doesn't matter. I don't know if my opinion is right. So I hope you can give me more clearer comments to point what should I change config files?

Why I need to do this, because test point is to verify that "some operations of double-list are running in constant time", I have saw the double-list code carefully, you are right, it can be verified by just reading the code. But it's not obvious and persuasive, so I think I need to write some codes to verify this.

andyross commented:
I don't know what CI policy is, and it's sort of in flux anyway (see the patch about "integration" mode).  @nashif would probably know best whether or not putting this in a benchmark is enough to satisfy unit test requirements, and would probably have different suggestions about where it should go.

tyrantlizard77 reviewed (changes requested):
Can we consider alternate approached to proving something is O(n) that does NOT use system time? If you want to know if a list is O(n) or a red-black tree is O(log2(n)) you count the number of operations and show that it is proportional as n increases.

NOT using the system time. This just breaks stuff in CI runs. We are already having problems with similar changes to the red-black tree test.

Zhaoningx commented:
Hi @andyross  @andrewboie ,I have updated my PR, would you like to review it again?Thanks.

And I don't know how to verify double list is O(1) without changing the original code. There is no interface API to count the search times. So I plan to remove this portion temporarily if there is no better method.

Zhaoningx commented:
The zephyrbot added a label "has conflicts", then, I used "wget https://github.com/zephyrproject-rtos/zephyr/pull/26570.patch",and apply the patch to the lasting master branch, but there is no conflict. Also as you see, there isn't a file marked of conflict.So I removed this label. @andrewboie @andyross

andrewboie commented:
> The zephyrbot added a label "has conflicts"

that seems to be in error. the PR says there are no rebase conflicts.

andrewboie commented:
> And I don't know how to verify double list is O(1) without changing the original code. There is no interface API to count the search times. So I plan to remove this portion temporarily if there is no better method.

If we have requirements that some mechanism has a particular big-O() asymptotic complexity, I am wondering what the guidance is from FUSA specs on this? We need to deal with this separately at any rate.

Zhaoningx commented:
> > And I don't know how to verify double list is O(1) without changing the original code. There is no interface API to count the search times. So I plan to remove this portion temporarily if there is no better method.
> 
> If we have requirements that some mechanism has a particular big-O() asymptotic complexity, I am wondering what the guidance is from FUSA specs on this? We need to deal with this separately at any rate.

We have requirements are "some operations like access to the "head" and "tail", insert and remove must run in constant time", so I think it's effective to proof the requirement if we can verify the operations of a double list are performing with O(1).

enjiamai commented on tests/benchmarks/data_structure_perf/dlist_perf/src/dlist_perf.c:
```diff
@@ -0,0 +1,177 @@
+ * @details
+ * Initialize a user defined struct contains dlist node
+ * Add into a doubly-linked list
+ * Enumerate the doubly-linked list.
```
Could you give it more descriptions about the test method? thanks!

Zhaoningx replied:
Hi @enjiamai , I have done this, thanks for your comments. :)

Zhaoningx commented on tests/benchmarks/data_structure_perf/dlist_perf/src/dlist_perf.c:
```diff
@@ -0,0 +1,184 @@
+ * @brief Test dlist for each function
+ *
+ * @details Initialize a double-linked list.
+ * And into a doubly-linked list and enumerate
```
Done, Thanks enjia. :)

LeiW000 reviewed (approved):
Approve it since Ning has already addressed the all comments.
